src/rcl_lifecycle_bindings.cpp

```diff
@@ -0,0 +1,363 @@
+// Copyright (c) 2020 Wayne Parrott. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rcl_lifecycle_bindings.hpp"
+
+#include <rcl/error_handling.h>
+#include <rcl_lifecycle/rcl_lifecycle.h>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "macros.hpp"
+#include "rcl_handle.hpp"
+#include "rcl_utilities.hpp"
+#include "lifecycle_msgs/msg/transition_event.h"
+#include "lifecycle_msgs/srv/change_state.h"
+#include "lifecycle_msgs/srv/get_available_states.h"
+#include "lifecycle_msgs/srv/get_available_transitions.h"
+#include "lifecycle_msgs/srv/get_state.h"
+
+
+namespace rclnodejs {
+
+static v8::Local<v8::Object> wrapState(const rcl_lifecycle_state_t* state) {
+  v8::Local<v8::Object> jsState = Nan::New<v8::Object>();
+  Nan::Set(jsState, Nan::New("id").ToLocalChecked(), Nan::New(state->id));
+  Nan::Set(jsState, Nan::New("label").ToLocalChecked(),
+           Nan::New(state->label).ToLocalChecked());
```

Comment:
My initial thought regarding your suggestion to bind JS objects to C++ object wrappers on rcl C structs is a more object-oriented approach and a departure from the current functional style addon api design. I'm hesitant to embark on a new approach without broader discussion and agreement. It seems like a slippery slope that could quickly get very complicated when generally applied.

I realize my impl dynamically creates JS data-transfer objects and copies data into them rather than directly referencing the C data from JS objects. In this case is seemed like an OK tradeoff of a little extra memory per state/transition data for a simple implementation. Maybe this is the wrong attitude given this code could be running on a smaller system, e.g., raspi. 

My preference is to continue with the current api design. But I'm open (team player) to a general change towards an object-oriented wrapping approach. Before we go there, consider the potential implementation complexity over our functional model. My expectation is adopting an OO approach will result in creating parallel C++ hierarchies wrapping rcl C structs and api with corresponding JS bindings. So in the lifecycle case that could involve creating C++ representations for Lifecycle StateMachine, State, Transition, TransitionDescription and Service.

@minggangw @koonpeng  Thoughts?

Reply:
I don't mind the current design, the only small problem I see is that some functions are creating a temporary object which are immediately discarded on the js side. But I don't think these functions are called often enough that it matters

Currently these are only simple plain data objects, but if we want to add methods to them in the future, it could create quite a lot of extra complexity that might not be worth the trouble.

Reply:
Agree. Thx for the feedback.